contractcourt/htlc_incoming_contest_resolver.go

```diff
@@ -206,71 +217,91 @@ func (h *htlcIncomingContestResolver) Resolve() (ContractResolver, error) {
-	case *invoices.HtlcSettleResolution:
-		return processHtlcResolution(resolution)
+		defer h.Registry.HodlUnsubscribeAll(hodlChan)
 
-	// If the resolution is nil, the htlc was neither settled nor failed so
-	// we cannot take action at present.
-	case nil:
+		// Take action based on the resolution we received. If the htlc
+		// was settled, or a htlc for a known invoice failed we can
+		// resolve it directly. If the resolution is nil, the htlc was
+		// neither accepted nor failed, so we cannot take action yet.
+		switch res := resolution.(type) {
+		case *invoices.HtlcFailResolution:
+			// In the case where the htlc failed, but the invoice
+			// was known to the registry, we can directly resolve
+			// the htlc.
+			if res.Outcome != invoices.ResultInvoiceNotFound {
+				return processHtlcResolution(resolution)
+			}
 
-	default:
-		return nil, fmt.Errorf("unknown htlc resolution type: %T",
-			resolution)
-	}
+		// If we settled the htlc, we can resolve it.
+		case *invoices.HtlcSettleResolution:
+			return processHtlcResolution(resolution)
 
-	// With the epochs and preimage subscriptions initialized, we'll query
-	// to see if we already know the preimage.
-	preimage, ok := h.PreimageDB.LookupPreimage(h.htlc.RHash)
-	if ok {
-		// If we do, then this means we can claim the HTLC!  However,
-		// we don't know how to ourselves, so we'll return our inner
-		// resolver which has the knowledge to do so.
-		if err := applyPreimage(preimage); err != nil {
-			return nil, err
+		// If the resolution is nil, the htlc was neither settled nor
+		// failed so we cannot take action at present.
+		case nil:
+
+		default:
+			return nil, fmt.Errorf("unknown htlc resolution type: %T",
+				resolution)
+		}
+	} else {
+		// If the HTLC hasn't expired yet, then we may still be able to
+		// claim it if we learn of the pre-image, so we'll subscribe to
+		// the preimage database to see if it turns up, or the HTLC
+		// times out.
+		//
+		// NOTE: This is done BEFORE opportunistically querying the db,
+		// to ensure the preimage can't be delivered between querying
+		// and registering for the preimage subscription.
+		preimageSubscription := h.PreimageDB.SubscribeUpdates()
+		defer preimageSubscription.CancelSubscription()
+
+		// With the epochs and preimage subscriptions initialized, we'll
+		// query to see if we already know the preimage.
+		preimage, ok := h.PreimageDB.LookupPreimage(h.htlc.RHash)
+		if ok {
+			// If we do, then this means we can claim the HTLC!
+			// However, we don't know how to ourselves, so we'll
+			// return our inner resolver which has the knowledge to
+			// do so.
+			if err := applyPreimage(preimage); err != nil {
+				return nil, err
+			}
+
+			return &h.htlcSuccessResolver, nil
 		}
 
-		return &h.htlcSuccessResolver, nil
+		witnessUpdates = preimageSubscription.WitnessUpdates
 	}
 
 	for {
-
 		select {
-		case preimage := <-preimageSubscription.WitnessUpdates:
+		case preimage := <-witnessUpdates:
```

Comment:
I miss writing of the success resolution outcome in this resolver. Shouldn't the happy path also be concluded with a write to the new bucket?

Reply:
The success resolver handles the write, since the claim has not confirmed yet at this stage.